### DISCUSSION TITLE
How should I port this to Linux if it uses WTL, which there's no documentation?

### DISCUSSION THREAD

**Discussion by ghost** [project64/project64#2044] [Support] [answered]
The last update was in 2015.

**Comment by project64** [comment, marked as answer]
the UI (Project64.exe) uses wtl, which is a wrapper around the windows api. You would need some type of new ui for linux.

**Reply by ghost** [reply]
Yeah I got that, but how should I switch to another UI if I can't understand WTL? How am I supposed to learn WTL if there is no documentation for it? @project64

**Reply by oddMLan** [reply]
honestly it's probably better to make a gui from scratch rather than trying to port the current GUI code
there's RMG https://github.com/Rosalie241/RMG which should be identical to PJ64's GUI tho that's adapted to mupen64plus, you could take its gui code and adapt it for PJ64

**Reply by ghost** [reply]
Probably, really sucks tho

**Reply by ghost** [reply]
For removing the current GUI and create a new one I need to learn WTL. Which sucks because there's no documentation

**Reply by oddMLan** [reply]
You don't actually. Just recreate the visual placement of the current UI using a different UI toolkit

**Reply by ghost** [reply]
@oddMLan 
Ok so for example,
How do I make the game visible? Do I need to change the renderer?
Is there a function to load roms so I can link it to a button or something?
It is hard when you didn't write the code and you can't constantly ask questions without a 15 min response time.

**Reply by bdragoncore** [reply]
I recommend to getting the core C++ source to compile first on linux before touching anything related with the UI (or even graphics). Build some simple command line app that hooks into the PJ64 core.

Loading a ROM is not simply a single function, there is a lot that goes on, like determining if the ROM should be byte-swapped, computing the CIC hash, and copying IPL portion into RSP DMEM, which also means you need the unified memory bus working before that.  

For N64 graphics, this is something handled by a plugin, and there are 2 approaches to the graphics emulation such as HLE and LLE.  

LLE RDP graphics means you need to rely on a LLE RSP plugin (the system's digital signal processor). The RSP is usually tasked by games to convert 3D resources into low-level RDP draw commands. 

HLE graphics gets pretty complicated too because it relies on the detected graphics microcode task via hashcodes, and it has handle the game's RSP microcodes in high-level C routines in order to capture and translate graphics commands/data into instructions for the host's graphics backend (eg: Direct3D, OpenGL, Vulkan, etc).

So before touching plugins, get the core CPU / memory / interfaces to work, look at Peter Lemon's unit tests (https://github.com/PeterLemon/N64), first you'd want to be able to get the CPU to be able to draw a simple bitmap into the video framebuffer.

I started out my emulator with just command line, getting the core to function based on unit testing, then attempting to draw pixels into the framebuffer. To display the framebuffer, I used a library such as GLFW that makes it quick to allocate a native GUI window resource + OpenGL context, I pushed the raw the RDRAM framebuffer to the GPU as a texture.

**Comment by project64** [comment]
maybe look at ImGUI and see  how that works, maybe look at https://github.com/stenzek/duckstation
